File: backend/resume_parser.py
import pdfplumber

SKILLS_DB = [
    "python",
    "java",
    "c++",
    "machine learning",
    "deep learning",
    "sql",
    "javascript",
    "react",
    "node",
    "tensorflow",
    "pytorch",
    "nlp",
    "data science",
    "flask",
    "fastapi"
]


def extract_text_from_pdf(file):

    text = ""

    with pdfplumber.open(file) as pdf:
        for page in pdf.pages:

            page_text = page.extract_text()

            if page_text:
                text += page_text

    return text.lower()


def extract_skills(text):
    if not text:
        return []

    found_skills = []

    for skill in SKILLS_DB:
        if skill in text:
            found_skills.append(skill)

    return list(set(found_skills))

File: backend/src/services/resumeParser.js
const pdfParse = require("pdf-parse");

const SKILLS_DB = [
  "python",
  "java",
  "c++",
  "machine learning",
  "deep learning",
  "sql",
  "javascript",
  "react",
  "node",
  "tensorflow",
  "pytorch",
  "nlp",
  "data science",
  "flask",
  "fastapi",
];

/**
 * Extract raw text from a PDF buffer.
 * @param {Buffer} buffer – the uploaded file buffer
 * @returns {Promise<string>}
 */
async function extractTextFromPdf(buffer) {
  const data = await pdfParse(buffer);
  return (data.text || "").toLowerCase();
}

/**
 * Match known skills inside the extracted text.
 * @param {string} text
 * @returns {string[]}
 */
function extractSkills(text) {
  if (!text) return [];

  const found = new Set();
  for (const skill of SKILLS_DB) {
    if (text.includes(skill)) {
      found.add(skill);
    }
  }
  return [...found];
}

module.exports = { extractTextFromPdf, extractSkills };
